# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Aaron Pelton

## Repo: https://github.com/hrawp/dm_nlp_week4

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

## Question 1 - Week 4 Project

In [46]:
import requests
import json

result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)
print(result)
with open("lyrics_to_json.json", "w", encoding="utf-8") as f:
    json.dump(result, f, indent=2, ensure_ascii=False)

{'lyrics': "I'm your only friend \nI'm not your only friend \nBut I'm a little glowing friend \nBut really I'm not actually your friend \nBut I am \n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI have a secret to tell \n\nFrom my electrical well \n\nIt's a simple message and I'm leaving out the whistles and bells \n\nSo the room must listen to me \n\nFilibuster vigilantly \n\nMy name is blue canary one note* spelled l-i-t-e \n\nMy story's infinite \n\nLike the Longines Symphonette it doesn't rest \n\n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI'm your only friend \n\nI'm not your only friend \n\

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

## Question 2 - Week 4 Project

In [ ]:
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

#  Load spaCy model
nlp = spacy.load("en_core_web_sm")

#  Add component by registered name (DO NOT instantiate it yourself!)
if "spacytextblob" not in nlp.pipe_names:
    nlp.add_pipe("spacytextblob", last=True)

#  Load lyrics from JSON file
with open("lyrics_to_json.json", "r", encoding="utf-8") as f:
    data = json.load(f)

print(data["lyrics"]) 

lyrics = data["lyrics"]

#  Run sentiment analysis
doc = nlp(lyrics)

#  Print sentiment scores
print("Polarity:", doc._.blob.polarity)
print("Subjectivity:", doc._.blob.subjectivity)

#  What I think
#  I think the polarity does not capture the negative nature of this song.  The whole concept of a light switch being or not being my friend makes it seem like I have no other friends.  I don't want to carve out a place in my soul for a light swtich.  That's very depressing.





I'm your only friend 
I'm not your only friend 
But I'm a little glowing friend 
But really I'm not actually your friend 
But I am 


Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I have a secret to tell 

From my electrical well 

It's a simple message and I'm leaving out the whistles and bells 

So the room must listen to me 

Filibuster vigilantly 

My name is blue canary one note* spelled l-i-t-e 

My story's infinite 

Like the Longines Symphonette it doesn't rest 



Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I'm your only friend 

I'm not your only friend 

But I'm a little glowing friend 

But really I'm not actually your f

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

## Question 3 - Week 4 Project

In [ ]:
import json
import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

def fetch_and_analyze_lyrics(band, song, filename):
    """
    Fetches lyrics from the lyrics.ovh API, saves them to a JSON file.

    Args:
        band (str): Name of the band or artist.
        song (str): Name of the song.
        filename (str): File path to save the lyrics JSON.

    Returns:
        dict: Dictionary with full lyrics text.
    """
    # Build API URL
    url = f"https://api.lyrics.ovh/v1/{band}/{song}"

    # Fetch lyrics
    response = requests.get(url)
    if response.status_code != 200:
        raise ValueError(f"Failed to fetch lyrics: {response.status_code} - {response.text}")

    result = json.loads(response.text)

    # Save to JSON file
    with open(filename, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2, ensure_ascii=False)

    # Output result
    # print("Lyrics:\n", result)  # Print first 500 chars


    return {
        "lyrics": result

    }


fetch_and_analyze_lyrics("The Beatles", "Yesterday", "yesterday.json")

{'lyrics': {'lyrics': "Yesterday\r\nAll my troubles seemed so far away\r\nNow it looks as though they're here to stay\r\nOh I believe in yesterday\r\nSuddenly\n\nI'm not half the man i used to be\n\nThere's a shadow hanging over me\n\nOh yesterday came suddenly\n\n\n\nWhy she had to go I don't know\n\nShe wouldn't say\n\nI said something wrong\n\nNow I long for yesterday\n\n\n\nYesterday\n\nLove was such an easy game to play\n\nNow I need a place to hide away\n\nOh I believe in yesterday\n\n\n\nWhy she had to go i don't know\n\nShe wouldn't say\n\nI said something wrong\n\nNow I long for yesterday\n\n\n\nYesterday\n\nLove was such an easy game to play\n\nNow i need a place to hide away\n\nOh i believe in yesterday\n\n\n\nMmm..."}}

4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

## Question 4 - Week 4 Project

In [ ]:
import json
import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

def analyze_lyrics(filename):
    """
    Rruns sentiment analysis using spaCy and spaCyTextBlob.

    Args:
        filename (str): File path to load the lyrics JSON.

    Returns:
        dict: Dictionary with polarity, subjectivity, and full lyrics text.
    """
    
    # Load spaCy model
    nlp = spacy.load("en_core_web_sm")

    # Add sentiment component if not already added
    if "spacytextblob" not in nlp.pipe_names:
        nlp.add_pipe("spacytextblob", last=True)

    # Load the saved lyrics
    with open(filename, "r", encoding="utf-8") as f:
        data = json.load(f)

    lyrics = data.get("lyrics", "")

    # Analyze sentiment
    doc = nlp(lyrics)

    # Output result
    print("Lyrics:\n", lyrics[:500], "...")  # Print first 500 chars
    print("\nTitle:", filename)
    print("\nSentiment:")
    print("Polarity:", doc._.blob.polarity)
    print("Subjectivity:", doc._.blob.subjectivity)

    return {
        "lyrics": lyrics,
        "polarity": doc._.blob.polarity,
        "subjectivity": doc._.blob.subjectivity
    }


analyze_lyrics("blackbird.json")
analyze_lyrics("cometogether.json")
analyze_lyrics("something.json")
analyze_lyrics("yesterday.json")

Lyrics:
 Blackbird singing in the dead of night
Take these broken wings and learn to fly
All your life
You were only waiting for this moment to arise.
Blackbird singing in the dead of night

Take these sunken eyes and learn to see

All your life

You were only waiting for this moment to be free.



Blackbird fly Blackbird fly

Into the light of the dark black night.



Blackbird fly Blackbird fly

Into the light of the dark black night.



Blackbird singing in the dead of night

Take these broken win ...

Title: blackbird.json

Sentiment:
Polarity: 0.17246376811594208
Subjectivity: 0.7072463768115943
Lyrics:
 Here come old flat top he come growing up slowly
He got joo-joo eyeball he one holy roller
He got hair down to his knees
Got to be a joker he just do what he please
He wear no shoeshine he got toe-jam football

He got monkey finger he shoot coca-cola

He say "I know you, you know me"

One thing I can tell you is you got to be free

Come together right now over me



He bag product

{'lyrics': "Yesterday\r\nAll my troubles seemed so far away\r\nNow it looks as though they're here to stay\r\nOh I believe in yesterday\r\nSuddenly\n\nI'm not half the man i used to be\n\nThere's a shadow hanging over me\n\nOh yesterday came suddenly\n\n\n\nWhy she had to go I don't know\n\nShe wouldn't say\n\nI said something wrong\n\nNow I long for yesterday\n\n\n\nYesterday\n\nLove was such an easy game to play\n\nNow I need a place to hide away\n\nOh I believe in yesterday\n\n\n\nWhy she had to go i don't know\n\nShe wouldn't say\n\nI said something wrong\n\nNow I long for yesterday\n\n\n\nYesterday\n\nLove was such an easy game to play\n\nNow i need a place to hide away\n\nOh i believe in yesterday\n\n\n\nMmm...",
 'polarity': 0.010000000000000002,
 'subjectivity': 0.5955555555555556}

## Analysis for Question 4
Blackbird:  I thought this was a good song and seemed positive.  I agree with the polarity score of 0.172.

Come Together: I always thought this song was positive, but looking at the lyrics as text I see some negativity.  I agree with the polarity score of 0.067.

Something: I thought this song was very positive and agree with a polarity scoree of 0.231.

Yesterday:  This for sure seemed more negative that the others.  That is reflected in the polarity score of almost 0 at 0.010.  I think the polarity could have gone negative though.

This was a very fun project.  It took me awhile, but I like the results.